core/services/postgres/advisory_lock.go

```diff
@@ -0,0 +1,119 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+// NOTE: All advisory lock class IDs used by the Chainlink application MUST be
+// kept here to avoid accidental re-use
+const (
+	AdvisoryLockClassID_EthBroadcaster int32 = 0
+	AdvisoryLockClassID_JobSpawner     int32 = 1
+	AdvisoryLockClassID_EthConfirmer   int32 = 2
+
+	// ORM takes lock on 1027321974924625846 which splits into ClassID 239192036, ObjID 2840971190
+	AdvisoryLockClassID_ORM int32 = 239192036
+
+	AdvisoryLockObjectID_EthConfirmer int32 = 0
+)
+
+type AdvisoryLock struct {
+	URI  string
+	conn *sql.Conn
+	db   *sql.DB
+	mu   *sync.Mutex
+}
+
+// TODO(sam): Fix this to be a single global connection not many individual ones
+// See: https://www.pivotaltracker.com/story/show/175288169
+func NewAdvisoryLock(uri string) *AdvisoryLock {
+	return &AdvisoryLock{
+		URI: uri,
+		mu:  &sync.Mutex{},
+	}
+}
+
+func (lock *AdvisoryLock) Close() error {
+	lock.mu.Lock()
+	defer lock.mu.Unlock()
+
+	var connErr, dbErr error
+
+	if lock.conn != nil {
+		connErr = lock.conn.Close()
+		if connErr == sql.ErrConnDone {
+			connErr = nil
+		}
+	}
+	if lock.db != nil {
+		dbErr = lock.db.Close()
+		if dbErr == sql.ErrConnDone {
+			dbErr = nil
+		}
+	}
+
+	lock.db = nil
+	lock.conn = nil
+
+	return multierr.Combine(connErr, dbErr)
+}
+
+func (lock *AdvisoryLock) TryLock(ctx context.Context, classID int32, objectID int32) (err error) {
+	lock.mu.Lock()
+	defer lock.mu.Unlock()
+	defer utils.WrapIfError(&err, "TryAdvisoryLock failed")
```

Comment:
I feel like the intention here is good but ends up obscuring the error in a way that literally writing out `errors.Wrap(err, "more context for this error at this particular line")` would, e.g:

```go

rows, err := lock.conn.QueryContext(ctx, "SELECT pg_try_advisory_lock($1, $2)", classID, objectID)
if err != nil {
  return errors.Wrap(err, "failed to select query lock")
}
```

Reply:
FWIW, `utils.WrapIfError` passes through to `errors.Wrap`, which annotates errors with a stack trace.  how Go handles `defer` statements, that stack trace points to the exact line at which the `return err` happens, meaning that we get the precise info that you're referring to.

Reply:
Yeah I'm referring more to the English part of the error message.

Reply:
We're going to be gutting this entire type very soon in favor of `postgres.AdvisoryLock`.  I'll make a note in the tracker story to take care of this as well.